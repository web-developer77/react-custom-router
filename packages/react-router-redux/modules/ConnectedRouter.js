import invariant from 'invariant'
import React, { Component, PropTypes } from 'react'
import { Router, Route } from 'react-router'

import { LOCATION_CHANGE } from './reducer'

class ConnectedRouter extends Component {
  static propTypes = {
    store: PropTypes.object,
    history: PropTypes.object,
    children: PropTypes.node
  }

  static contextTypes = {
    store: PropTypes.object
  }

  componentWillMount() {
    const { children } = this.props

    invariant(
      children == null || React.Children.count(children) === 1,
      'A <ConnectedRouter> may have only one child element'
    )
  }

  render() {
    const { store:propsStore, history, children, ...props } = this.props
    let store = propsStore || this.context.store

    return (
      <Router {...props} history={history}>
        <Route render={({ location }) => {
            store.dispatch({
              type: LOCATION_CHANGE,
              payload: location
            })

            return children ? React.Children.only(children) : null
          }}/>
      </Router>
    )
  }
}

export default ConnectedRouter
