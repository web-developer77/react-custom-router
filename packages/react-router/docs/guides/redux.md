# Redux Integration

Redux is an important part of the React ecosystem. We want to make the integration of React Router and Redux as seamless as possible for people wanting to use both.

## Blocked Updates

Generally, React Router and Redux work just fine together. Occasionally though, an app can have a component that doesn't update when the location changes (child routes or active nav links don't update).

This happens if:

1. The component is connected to redux via `connect()(Comp)`.
2. The component is **not** a "route component", meaning it is not
   rendered like so: `<Route component={SomeConnectedThing}/>`

The problem is that Redux implements `shouldComponentUpdate` and there's no indication that anything has changed if it isn't receiving props from the router.  This is straightforward to fix. Find where you `connect` your component and wrap it in `withRouter`.

```js
// before
export default connect(mapStateToProps)(Something)

// after
import { withRouter } from 'react-router-dom'
export default withRouter(connect(mapStateToProps)(Something))
```

## Deep integration

Some folks want to:

- synchronize the routing data with, and accessed from, the store
- be able to navigate by dispatching actions
- have support for time travel debugging for route changes in the Redux
  devtools

All of this requires deeper integration. Please note you don't need this deep integration:

- Route changes are unlikely to matter for time travel debugging.
- Rather than dispatching actions to navigate you can pass the `history` object provided to route components to your actions and navigate with it there.
- Routing data is already a prop of most of your components that care about it, whether it comes from the store or the router doesn't change your component's code.

However, we know some people feel strongly about this and so we want to provide the best deep integration possible. As of version 4 of React Router, the React Router Redux package is a part of the project.  Please refer to it for deep integration.

[React Router Redux](https://github.com/reacttraining/react-router/tree/master/packages/react-router-redux)

