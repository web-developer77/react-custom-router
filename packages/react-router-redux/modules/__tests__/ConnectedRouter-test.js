import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, combineReducers } from 'redux'
import { Provider } from 'react-redux'
import createHistory from 'history/createMemoryHistory'

import ConnectedRouter from '../ConnectedRouter'
import { routerReducer } from '../reducer'

describe('A <ConnectedRouter>', () => {
  let store, history

  beforeEach(() => {
    store = createStore(combineReducers({
      router: routerReducer
    }))

    history = createHistory()
  })

  it('connects to a store via Provider', () => {
    expect(store.getState()).toHaveProperty('router.location', null)

    renderer.create(
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <div>Test</div>
        </ConnectedRouter>
      </Provider>
    )

    expect(store.getState()).toHaveProperty('router.location.pathname')
  })

  it('connects to a store via props', () => {
    expect(store.getState()).toHaveProperty('router.location', null)

    renderer.create(
      <ConnectedRouter store={store} history={history}>
        <div>Test</div>
      </ConnectedRouter>
    )

    expect(store.getState()).toHaveProperty('router.location.pathname')
  })

  it('updates the store with location changes', () => {
    renderer.create(
      <ConnectedRouter store={store} history={history}>
        <div>Test</div>
      </ConnectedRouter>
    )

    expect(store.getState()).toHaveProperty('router.location.pathname', '/')

    history.push('/foo')

    expect(store.getState()).toHaveProperty('router.location.pathname', '/foo')
  })

  describe('with children', () => {
    it('to render', () => {      
      const tree = renderer.create(
        <ConnectedRouter store={store} history={history}>
          <div>Test</div>
        </ConnectedRouter>
      ).toJSON()
      
      expect(tree).toMatchSnapshot()
    })
  })

  describe('with no children', () => {
    it('to render', () => {      
      const tree = renderer.create(
        <ConnectedRouter store={store} history={history} />
      ).toJSON()
      
      expect(tree).toMatchSnapshot()
    })
  })
})
