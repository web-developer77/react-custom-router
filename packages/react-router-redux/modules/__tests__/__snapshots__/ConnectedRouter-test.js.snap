// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A <ConnectedRouter> with children to render 1`] = `
<div>
  Test
</div>
`;

exports[`A <ConnectedRouter> with no children to render 1`] = `null`;
